content/gocourse-2019-05-melbourne.slide

```diff
@@ -813,50 +816,103 @@ Embedding types introduces the possiblity for naming conflicts. Go follows some
 .caption J. D. Frazer "Illiad", User Friendly - [[https://en.wikipedia.org/wiki/User_Friendly#Crud_Puppy][Crud Puppy]]
 
 
-* Errors (1/3)
+* Errors in Go
 
-"Errors are values."  - [[https://go-proverbs.github.io/][Rob Pike]]
+- `error` a builtin type. ([[https://golang.org/src/builtin/builtin.go?s=11246:11286#L251][source: error]])
+- Check errors explicitly
+- The [[https://golang.org/pkg/errors/][`errors`]] package has functions for creating new error type.
+- Creation of new error type can be done in a few ways:
 
-The `error` type is a built-in interface:
+	func Sqrt(f float64) (float64, error) {
+		if f < 0 {
+			return 0, errors.New("math: square root of negative number")
+		}
+		// implementation
+	}
+- Or use formatted error message using `fmt.Errorf`
+	if f < 0 {
+		return 0, fmt.Errorf("math: square root of negative number %g", f)
+	}
+
+* Errors in Go (...cotd)
 
-	type error interface {
-	    Error() string
+- Custom error types can be created. Example from `net` package -
+
+	type Error interface {
+		error
+		Timeout() bool
+		Temporary() bool
 	}
+- Implementation of the `Error` interface here - [[https://golang.org/pkg/net/#AddrError][net/AddrError]]
+- The error values can be logged and printed. Example -
 
-By convention, functions often return `error` as the second/last return value.
-The calling code should handle errors by testing whether the error equals nil:
+	g, err := Sqrt(-1); err != nil {
+		log.Println(err)
+	}
 
-	func getCat(strID string) (*Cat, error) {
-		id, err := strconv.Atoi(strID)
-		if err != nil {
-		    return nil, CatError(err)
-		}
-		return catStore.cats[id], nil
+
+* Error Handling
+
+- Error handling: Check for the presence of an error (compare with nil).
+- Do not compare error variables. ([[https://goplay.space/#mgw9QAmeLUP][Example]])
+- Errors are values so it can be programmed. ([[https://goplay.space/#MTqqxwEdq-A][Example]])
+	g, err := Sqrt(-1); err != nil {
+		log.Println(err)
 	}
+- Handle errors gracefully ([[https://goplay.space/#kpFwYYBE0ND][A useful pattern]])
+- To check if an `error` is of certain type use custom error types with type assertion
+- Example: ([[https://goplay.space/#t4s_BrAQTLt][Errors and type assertion]])
+- By convention many standard library functions return `error` as the second return value.
 
+"Errors are values."  - [[https://go-proverbs.github.io/][Rob Pike]]
+
+* Defer
+
+- `defer` (keyword) allows you to postpone execution.
+- When `defer` statement executes, "the parameters to call are evaluated and saved anew"
+- Deferred functions are called when "function returns" or "stack unwinds".
```

Comment:
Why the double-quotes? The following would read more naturally:

> - When `defer` statement executes, the parameters to call are evaluated and saved anew
> - Deferred functions are called when the function returns or the stack unwinds.

If these are direct quotes, link to the source.

Reply:
I copied a part of this text from the refspec. But I think you're right, I'll remove the quotes.